internal/util/function/function_executor.go

```diff
@@ -0,0 +1,212 @@
+/*
+ * # Licensed to the LF AI & Data foundation under one
+ * # or more contributor license agreements. See the NOTICE file
+ * # distributed with this work for additional information
+ * # regarding copyright ownership. The ASF licenses this file
+ * # to you under the Apache License, Version 2.0 (the
+ * # "License"); you may not use this file except in compliance
+ * # with the License. You may obtain a copy of the License at
+ * #
+ * #     http://www.apache.org/licenses/LICENSE-2.0
+ * #
+ * # Unless required by applicable law or agreed to in writing, software
+ * # distributed under the License is distributed on an "AS IS" BASIS,
+ * # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * # See the License for the specific language governing permissions and
+ * # limitations under the License.
+ */
+
+package function
+
+import (
+	"fmt"
+	"sync"
+
+	"google.golang.org/protobuf/proto"
+
+	"github.com/milvus-io/milvus-proto/go-api/v2/commonpb"
+	"github.com/milvus-io/milvus-proto/go-api/v2/schemapb"
+	"github.com/milvus-io/milvus/internal/proto/internalpb"
+	"github.com/milvus-io/milvus/pkg/mq/msgstream"
+	"github.com/milvus-io/milvus/pkg/util/merr"
+)
+
+type Runner interface {
+	GetSchema() *schemapb.FunctionSchema
+	GetOutputFields() []*schemapb.FieldSchema
+
+	MaxBatch() int
+	ProcessInsert(inputs []*schemapb.FieldData) ([]*schemapb.FieldData, error)
+	ProcessSearch(placeholderGroup *commonpb.PlaceholderGroup) (*commonpb.PlaceholderGroup, error)
+}
+
+type FunctionExecutor struct {
+	runners map[int64]Runner
+}
+
+func createFunction(coll *schemapb.CollectionSchema, schema *schemapb.FunctionSchema) (Runner, error) {
+	switch schema.GetType() {
+	case schemapb.FunctionType_BM25: // ignore bm25 function
+		return nil, nil
+	case schemapb.FunctionType_OpenAIEmbedding:
```

Comment:
actually I am a bit concerned whether we should create a new FunctionType just for OpenAI? can we pass in OpenAI in one of the function params instead?

the problem with the current approach is: if we added a new embedding service provider in a Milvus patch version, the user will have to update their SDK version as well.

What I suggest:

```
case schemapb.FunctionType_ExternalEmbeddingService: // maybe choose a better name
    f, err := NewExternalEmbeddingServiceFunction(coll, schema)
    ...

func NewExternalEmbeddingServiceFunction(coll, schema) {
    func_params := schema.params;
    switch func_params["provider"] {
        case: "OpenAI":
            f, err := NewOpenAIEmbeddingFunction(coll, schema)
            if err { ... }
            return f
    }
}
```

Reply:
I have also considered this design. The disadvantage is that it may not be easy for users to fill in these parameters correctly. For example, the provider is openai or OpenAI or something similar. 
If other providers are integrated later, it might be difficult for users to fill in the information correctly without referring to the documentation.
We still need to add some helper functions to the SDK to assist users in filling out the parameters correctly.

Reply:
`ExternalEmbeddingService` is too broad, and it's difficult to determine the inputs and outputs. It can be changed to more specific names like `TextEmbedding`, `ImageEmbedding`, etc., with clearly defined inputs and outputs. This would make it easier for the SDK to perform validations. @zhengbuqian